pkg/hubble/recorder/sink/dispatch.go

```diff
@@ -0,0 +1,231 @@
+	d.sinkByRuleID[ruleID] = s
+	return &Handle{
+		C:    s.trigger,
+		sink: s,
+	}, nil
+}
+
+// UnregisterSink will stop and unregister the sink for the given ruleID.
+// It waits for any pending packets to be forwarded to the sink before closing
+// it and returns the final statistics or an error, if an error occurred.
+func (d *Dispatch) UnregisterSink(ctx context.Context, ruleID uint16) (stats Statistics, err error) {
+	d.mutex.Lock()
+	s, ok := d.sinkByRuleID[ruleID]
+	delete(d.sinkByRuleID, ruleID)
+	// unlock early to avoid holding the lock during s.close() which may block
+	d.mutex.Unlock()
+
+	if !ok {
+		return Statistics{}, fmt.Errorf("no sink found for rule id %d", ruleID)
+	}
+
+	if err = s.close(ctx); err != nil {
+		return Statistics{}, err
+	}
+
+	return s.copyStats(), nil
+}
+
+func (d *Dispatch) decodeRecordCaptureLocked(data []byte) (rec record, err error) {
+	dataLen := uint32(len(data))
+	if dataLen < monitor.RecorderCaptureLen {
+		return record{}, fmt.Errorf("not enough data to decode capture message: %d", dataLen)
+	}
+
+	// This needs to stay in sync with struct capture_msg from
+	// bpf/include/pcap.h.
+	// We could use binary.Read on monitor.RecorderCapture, but since it
+	// requires reflection, it is too slow to use on the critical path here.
+	const (
+		offsetRuleID         = 2
+		offsetBootTime       = 8
+		offsetCaptureLength  = 16
+		offsetOriginalLength = 20
+	)
+	n := byteorder.Native
+	ruleID := n.Uint16(data[offsetRuleID:])
+	bootTime := n.Uint64(data[offsetBootTime:])
+	capLen := n.Uint32(data[offsetCaptureLength:])
+	origLen := n.Uint32(data[offsetOriginalLength:])
+
+	// data may contain trailing garbage from the perf ring buffer
+	// https://lore.kernel.org/patchwork/patch/1244339/
+	packetEnd := monitor.RecorderCaptureLen + capLen
+	if dataLen < packetEnd {
+		return record{}, fmt.Errorf("capture record too short: want:%d < got:%d", dataLen, packetEnd)
+	}
+	packet := data[monitor.RecorderCaptureLen:packetEnd]
+
+	return record{
+		timestamp: d.bootTimeToWallTimeLocked(int64(bootTime)),
+		ruleID:    ruleID,
+		inclLen:   capLen,
+		origLen:   origLen,
+		data:      packet,
+	}, nil
+}
+
+func (d *Dispatch) bootTimeToWallTimeLocked(bootTime int64) time.Time {
+	elapsedSinceStart := time.Duration(bootTime - d.startBootTime)
+	return d.startWallTime.Add(elapsedSinceStart)
+}
+
+func getTimeNow() (bootTime int64, wallTime time.Time, err error) {
+	var bootTimespec unix.Timespec
+
+	// Ideally we would use __vdso_clock_gettime for both clocks here, to have
+	// as little overhead as possible such that the two timestamps are taken
+	// as close together as possible. time.Now() will actually use VDSO on
+	// Go 1.9+, but the unix.ClockGettime call is a regular system call for
+	// now.
+	wallTime = time.Now()
+	err = unix.ClockGettime(unix.CLOCK_BOOTTIME, &bootTimespec)
+	if err != nil {
+		return 0, time.Time{}, err
+	}
+
+	return bootTimespec.Nano(), wallTime, nil
```

Comment:
Why not return a single offset (elapsedSinceBoot) value here? This can be a `time.Duration`, since it's not suscept to time zone changes. Calling `time.Duration(bootTime - d.startBootTime)` in `bootTimeToWallTimeLocked()` is not necessary as far as I can see.

Reply:
You are right. I originally stored `wallTime` here to deal with DST changes and such, but since we're actually only interested in UTC (which does not have DST), returning a single offset will work (except when it comes to leap seconds, but I don't think `time.Time` handles them)